rasa/utils/tensorflow/models.py

```diff
@@ -211,11 +229,40 @@ def fit(
                         )
 
                     val_results = self._get_metric_results(prefix="val_")
+                    improved = self._update_best_metrics_so_far(val_results)
+                    if improved and self.best_model_file is not None:
+                        logger.debug(f"Creating model checkpoint at epoch={epoch}...")
+                        best_model_epoch = epoch
+                        self.save(self.best_model_file, overwrite=True)
 
                 postfix_dict.update(val_results)
 
             progress_bar.set_postfix(postfix_dict)
 
+        # Checkpoint the model one last time after training
+        if evaluate_on_num_examples > 0 and self.best_model_file is not None:
+            epoch_batch_size = self.linearly_increasing_batch_size(
+                epochs, batch_size, epochs
+            )
+            self._batch_loop(
+                evaluation_dataset_function,
+                tf_evaluation_on_batch_function,
+                epoch_batch_size,
+                False,
+                training_steps,
+                self.test_summary_writer,
+            )
+
+            val_results = self._get_metric_results(prefix="val_")
```

Comment:
We could extract a separate method for the following lines as we are using it in two places.

Reply:
Do you mean an alternative method for `_get_metric_results()`?

Reply:
No, I mean we could extract 
```
                    if self.checkpoint_model and self._does_model_improve(val_results):
                        logger.debug(f"Creating model checkpoint at epoch={epoch}...")
                        best_model_epoch = epoch
                        self.save(self.best_model_file, overwrite=True)
```
into a method. And pass `val_result` as well as `epoch` to it.